pkg/provisioner/crdprovisioner.go

```diff
@@ -250,58 +290,68 @@ func (c *CrdProvisioner) PublishVolume(
 
 	attachmentName := azureutils.GetAzVolumeAttachmentName(volumeName, nodeID)
 
-	azVolumeAttachmentInstance, err := azVA.Get(ctx, attachmentName, metav1.GetOptions{})
+	if volumeContext == nil {
+		volumeContext = map[string]string{}
+	}
+	azVolumeAttachmentInstance, err := azVA.Get(ctx, strings.ToLower(attachmentName), metav1.GetOptions{})
 	if err == nil {
-		if azVolumeAttachmentInstance.Status != nil && azVolumeAttachmentInstance.Status.Error != nil {
-			azVolumeAttachmentInstance.Status = nil
-			azVolumeAttachmentUpdated, err := azVA.UpdateStatus(ctx, azVolumeAttachmentInstance, metav1.UpdateOptions{})
-			if err != nil || azVolumeAttachmentUpdated == nil || azVolumeAttachmentUpdated.Status != nil {
-				klog.Errorf("Failed to update azVolumeAttachment resource for volume name (%s), error: %v", volumeName, err)
+		updated := azVolumeAttachmentInstance.DeepCopy()
+		updated.Spec.RequestedRole = v1alpha1.PrimaryRole
+		if updated.Status != nil && updated.Status.Error != nil {
+			updated.Status = nil
+			updated, err = azVA.UpdateStatus(ctx, updated, metav1.UpdateOptions{})
+			if err != nil || updated == nil || updated.Status != nil {
+				klog.Errorf("Failed to update azVolumeAttachment resource status for volume name (%s), error: %v", volumeName, err)
 				return nil, err
 			}
-		} else {
+		}
 
-			// If there exists an attachment, we are trying to update
-			// the AzVolumeAttachment role from Replica to Primary
-			azVolumeAttachmentInstance.Spec.RequestedRole = v1alpha1.PrimaryRole
-			_, err := azVA.Update(ctx, azVolumeAttachmentInstance, metav1.UpdateOptions{})
-			if err != nil {
-				return nil, err
-			}
-			return azVolumeAttachmentInstance.Status.PublishContext, nil
+		// If there exists an attachment, we are trying to update
+		// the AzVolumeAttachment role from Replica to Primary
+		updated.Spec.RequestedRole = v1alpha1.PrimaryRole
+		_, err := azVA.Update(ctx, updated, metav1.UpdateOptions{})
+		if err != nil {
+			return nil, err
 		}
-	}
+		return updated.Status.PublishContext, nil
+	} else {
```

Comment:
@sunpa93 @nearora-msft I think this else was supposed to be on line 307?

Reply:
No. I think the merge has broken the logic fixed over here by @sunpa93 . I am fixing and pushing it, @sunpa93  Please review

Reply:
else is supposed to be there because we only should create a new object if there already wasn't one generated by the AzVolumeAttachment controller. The problem here is the `return` statement in line 316. It should reach the `wait.PollIntermediate` at line 363.